packages/devtools/lib/src/timeline/timeline_protocol.dart

```diff
@@ -112,102 +210,149 @@ class TimelineData {
-      // PipelineConsume event flows, and we want to consider each
-      // PipelineConsume flow to be its own event. This event can also parent
-      // irrelevant events that we do not want to track.
-      else if (!event.name.contains('MessageLoop::RunExpiredTasks')) {
-        _gpuEventNode = e;
-      }
+    if (current != null) {
+      current.addChild(timelineEvent);
     }
+    currentEventNodes[event.type.index] = timelineEvent;
   }
 
   void _handleDurationEndEvent(TraceEvent event) {
-    TimelineEvent current;
-    if (event.isCpuEvent && _cpuEventNode != null) {
-      _cpuEventNode.endTime = event.timestampMicros;
-      current = _cpuEventNode;
-
-      // Since this event is complete, move back up the stack.
-      _cpuEventNode = _cpuEventNode.parent;
-      if (_cpuEventNode == null) {
-        _maybeAddEvent(current);
+    TimelineEvent current = currentEventNodes[event.type.index];
+
+    if (current == null) return;
+
+    // If the names of [event] and [current] do not match, our event nesting is
+    // off balance due to duplicate events from the engine. Balance the tree so
+    // we can continue processing trace events for [current].
+    if (event.name != current.name) {
+      if (event.json.toString() ==
+          previousDurationEndEvents[event.type.index]?.json.toString()) {
+        // This is a duplicate of the previous DurationEnd event we received.
+        //
+        // Trace example:
+        // VSYNC - DurationBegin
+        // Framework Workload - DurationBegin
+        // ...
+        // FrameWork Workload - DurationEnd [previousDurationEndEvent]
+        // FrameWork Workload - DurationEnd ([event] - duplicate)
+        // VSYNC - DurationEnd
+        //
+        return;
+      } else if (current.name ==
+              previousDurationEndEvents[event.type.index]?.name &&
+          current.parent?.name == event.name &&
+          current.children.length == 1 &&
+          current.eventTraces.first.toString() ==
+              current.children.first.eventTraces.first.toString()) {
+        // There was a duplicate DurationBegin event associated with
+        // [previousDurationEndEvent]. [event] is actually the DurationEnd event
+        // for [current.parent]. Trim the extra layer created by the duplicate.
+        //
+        // Trace example:
+        // VSYNC - DurationBegin
+        // Framework Workload - DurationBegin (duplicate - remove this node)
+        // Framework Workload - DurationBegin
+        // ...
+        // FrameWork Workload - DurationEnd [previousDurationEndEvent]
+        // VSYNC - DurationEnd [event]
+        //
+        current.parent.removeChild(current);
+        current = current.parent;
+        currentEventNodes[event.type.index] = current;
+      } else {
+        // The current event node has fallen into an unrecoverable state. Reset
+        // the tracking node.
+        currentEventNodes[event.type.index] = null;
```

Comment:
when does this case occur? this sort of logic suggests we are not doing things in a robust enough way and we are in danger of displaying the wrong data to users.